### ISSUE TITLE
Can I mount longhorn volume as readOnly?

### ISSUE THREAD

**Issue by ElisaMeng** [longhorn/longhorn#1927]
When doing so, longhorn report error: 
`  Warning  FailedMount  4m7s (x699 over 23h)  kubelet  MountVolume.SetUp failed for volume "pvc-eef719cd-6e2f-42fd-80bd-ce400675c6c5" : rpc error: code = FailedPrecondition desc = Not support readOnly
`

Is this expected behaviour by design?

**Comment by yasker** [comment]
You should be able to (if you meant to mount it as read-only on one node).

What's the reproduce steps?

**Comment by joshimoo** [comment]
We currently error out in the Node.PublishVolume call when a read only access mode is requested.
https://github.com/longhorn/longhorn-manager/blob/master/csi/node_server.go#L77

I think we can support it, but we have to decide what we want to do, since even a regular read only mount, could still lead to writes to the disk for ext4 if there is a journal. This post describes this problem and links to the man pages.
https://askubuntu.com/a/830240

We can mount with `ro,noload` to prevent the journal replay, but then the fs might be dirty if there was a previous unclean mount.

Assuming we want to do this at the block dev level, than there is the problem of if we set the blockdev to read only that this only prevents io from userspace, see
https://lkml.org/lkml/2007/2/6/1

Not sure if that has been addressed in the meantime and would require further investigation.

**Comment by joshimoo** [comment]
TLDR, if we want to enable it I think we could go with the `ro,noload` for ext4 filesystems, but we should make a conscious choice to determine if that is sufficient guarantee for Read-only access. Since it's a filesystem mount, I would say it's sufficient to guarantee read only on the filesystem level (ie readonly mount with no load for the journal).

@yasker What do you think?

**Comment by yasker** [comment]
@ElisaMeng What's your use case? The implementation would be a bit awkward as @joshimoo described. Also, you're trying to use read-only volume on one node, not multiple nodes (ROX/ReadManyOnly), right?

**Comment by ElisaMeng** [comment]
I just want to mount the volume which is detached to examine its fs when I get data corruption, to make sure the they are not modify with new mount, I set it to readonly, and then endup in this situation.  Maybe I can do this with snapshot or copy instead?

**Comment by joshimoo** [comment]
@ElisaMeng for now to inspect the volume you can manually attach it to a node via the longhorn-ui, attach in `maintenance mode` there is checkbox for that. Then ssh into the node or run a privileged pod with full host access to /dev/longhorn.

Now you can manually mount the filesystem as read-only for inspection. Make sure you are aware of the `ro` & `noload` options, below is a link some information in regard to read only mounts of ext4 filesystems. 
https://www.sans.org/blog/how-to-mount-dirty-ext4-file-systems/

**Comment by ElisaMeng** [comment]
@joshimoo Thanks a lot, it is really helpful link!
